### PR TITLE
validation changes for assisted bare metal deployments

### PR DESCRIPTION
**Pull request by dhellmann**
The assisted installer is currently bypassing the installer validation
checks for the `baremetal` platform by using artificial data or
workarounds. This enhancement describes how we can modify the
installer to avoid those workarounds while retaining all of the checks
needed for the installer-provisioned infrastructure workflow.

### CODE REVIEW

**Comment by stbenjam** [comment]
Thanks for writing this up. I think it captures all the discussions and alternative proposals. 

By doing the minimal set of validations at the InstallConfig level you can produce manifests and ignitions without needing the bootstrap infrastructure available. There’s probably some benefits there by separating the concerns even without the assisted install case.

File: enhancements/installer/assisted-installer-bare-metal-validations.md [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+  or installer-provisioned infrastructure installer.
+- Removing or relaxing any validation rules for the
+  installer-provisioned infrastructure installer.
+- Recapitulate the test plans for the assisted installer.
+
+## Proposal
+
+All of the validation is currently performed during early phases of
+the installer for the `InstallConfig` target. The assisted installer
+does not phase that for the `Cluster` target, so if we move some of
```
Comment (stbenjam) [inline enhancements/installer/assisted-installer-bare-metal-validations.md:63]: Nit: Is a verb missing before phase?

**Comment by yevgeny-shnaidman** [comment]
@dhellmann doesn't it look a little bit strange that for some platforms we enforce validations during InstallConfig loading, and for some only during Cluster generation?

**Comment by stbenjam** [comment]
Why do you think it’s strange? The provision and platform checks that already exist are a better fit for the kind of checks we are doing.

**Comment by romfreiman** [comment]
I'm fine with this approach (as I stated multiple times already). My only concern is whether we're not breaking API here.
I mean if now someone expect to fail during ignition phase (maybe there are even tests that are verifying this?) but will fail later on. Would it affect VZN? 
Unfortunately I dont have this knowledge, hence raising the concern

**Comment by stbenjam** [comment]
I don't think any IPI user will care, and it's probably more correct and consistent with other platforms with this approach. You'd be able to run create manifests and ignition-configs from your laptop, and then bring the result to where you intend to provision, for example.

No one may even notice the change for the IPI workflow, a good percentage of users just run create cluster directly, and moving these validations later will be imperceptible (milliseconds).

File: enhancements/installer/assisted-installer-bare-metal-validations.md [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+[PlatformProvisionCheck](https://github.com/openshift/installer/blob/master/pkg/asset/installconfig/platformprovisioncheck.go)
+and
+[PlatformCredsCheck](https://github.com/openshift/installer/blob/master/pkg/asset/installconfig/platformcredscheck.go)
+assets are both dependencies for the `Cluster` asset, which means they
+are evaluated later in the installation process. Both already have
+platform-specific logic, although not for the `baremetal` platform.
+
+The assisted installer does not need the `bootstrapProvisioningIP`,
+`bootstrapProvisioningIP`, `provisioningNetworkCIDR`,
+`provisioningDHCPRange`, or `provisioningNetworkInterface` fields. The
```
Comment (stbenjam) [inline enhancements/installer/assisted-installer-bare-metal-validations.md:109]: What do you think about clusterOSImage and bootstrapOSImage? We could validate that if specified (they're not in the assisted install), that it's a valid URL in the installConfig, and only validate it exists in ProvisionCheck.

File: enhancements/installer/assisted-installer-bare-metal-validations.md (hunk 1 above)
Comment (dhellmann) [reply enhancements/installer/assisted-installer-bare-metal-validations.md:63]: Replaced word salad with English.

File: enhancements/installer/assisted-installer-bare-metal-validations.md [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+The
+[PlatformProvisionCheck](https://github.com/openshift/installer/blob/master/pkg/asset/installconfig/platformprovisioncheck.go)
+and
+[PlatformCredsCheck](https://github.com/openshift/installer/blob/master/pkg/asset/installconfig/platformcredscheck.go)
+assets are both dependencies for the `Cluster` asset, which means they
+are evaluated later in the installation process. Both already have
+platform-specific logic, although not for the `baremetal` platform.
+
+The assisted installer does not need the `bootstrapProvisioningIP`,
+`bootstrapProvisioningIP`, `provisioningNetworkCIDR`,
```
[the remark this answers is not in the document]
Comment (dhellmann) [reply enhancements/installer/assisted-installer-bare-metal-validations.md:108]: fixed

File: enhancements/installer/assisted-installer-bare-metal-validations.md (hunk 2 above)
Comment (dhellmann) [reply enhancements/installer/assisted-installer-bare-metal-validations.md:109]: I added a paragraph about splitting those checks.

File: enhancements/installer/assisted-installer-bare-metal-validations.md [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+ensure the URL exists. Those checks can be split between the two
+phases, so that the URL format validation is done in
+`ValidateInstallConfig()` and the existence check is moved to
+`PlatformProvisionCheck`.
+
+The assisted installer does not need the `libvirtURI` or host BMC
+credentials, so the checks for those values could move to a new public
+function in the `pkg/types/baremetal/validation` package that can be
+invoked from
+[PlatformCredsCheck.Generate()](https://github.com/openshift/installer/blob/master/pkg/asset/installconfig/platformcredscheck.go#L60)
```
Comment (stbenjam) [inline enhancements/installer/assisted-installer-bare-metal-validations.md:127]: Should we link to the upstream work to allow BMC credentials to be empty?

https://github.com/metal3-io/baremetal-operator/pull/563

**Comment by stbenjam** [comment]
Thanks, I'm happy with this.

@hardys is out for a couple days, @crawford could you have a look at the proposal?

/assign @crawford

File: enhancements/installer/assisted-installer-bare-metal-validations.md [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+validation rules to skip. There are two reasons not to do this. First,
+it would expose the validation rules through the installer's public
+API, even though we would only want the assisted installer to use the
+feature. Second, it places the burden of managing the set of
+validations on the wrapper in the assisted installer, instead of on
+the author of validation rules in the installer source code. We know
+which validations should apply in each case, so there is no need to
+make the validations configurable.
+
+### Add an environment variable
```
Comment (kirankt) [inline enhancements/installer/assisted-installer-bare-metal-validations.md:220]: Pretty sure this is a non-starter. The installer team will reject this idea based on previous experience.

File: enhancements/installer/assisted-installer-bare-metal-validations.md (hunk 4 above)
Comment (dhellmann) [reply enhancements/installer/assisted-installer-bare-metal-validations.md]: added

File: enhancements/installer/assisted-installer-bare-metal-validations.md (hunk 5 above)
Comment (dhellmann) [reply enhancements/installer/assisted-installer-bare-metal-validations.md:226]: I agree, that's why it's down here in the alternatives list.

**Comment by markmc** [comment]
I think we're missing a description of how exactly openshift-install will be used in an assisted installation. It's touched on in this enhancement and #376, but I think anyone not close to the subject would come away quite confused.

Like, firstly an assisted install only ever uses "create ignition-configs" and not "create cluster". (I guess some would say this implies it's a UPI use case)

But rather than the typical `platform: none` for UPI, we want to use `platform: baremetal` to express some intent about how the cluster should be installed, and its final configuration.

My understanding of that intent is: (a) configuring the keepalived/haproxy API load balancer (which is IPI-ish), and (b) enable the baremetal-operator for day 2 machine API operations (usually only available for IPI). And not (c) "provision a libvirt VM with the bootstrap configuration, and a bunch of IPMI-controlled physical machines as control plane nodes using PXE" (definitely IPI).

For each `baremetal.Platform` field, can we make it clear which of those 3 capabilities the field relates to?

If we did, then this proposal becomes "a UPI-like use case for 2 of these 3 capabilities" and it would be more obvious which fields are required for that.

Related, AIUI the "configure the baremetal-operator" intent currently requires all of the fields in the `metal3.io Provisioning`? But we desire that to be a day 2 "set up PXE-based machine provisioning" thing? (That topic is touched on in openshift/machine-api-operator#560)

Describing this in terms of which validations run at which phase doesn't really resonate with me - it's more which `platform: baremetal` capabilities (of the three) are being requested, and which validations apply to those capabilities?

Maybe this enhancement could become more "allowing bare-metal UPI installs to use some baremetal IPI capabilities"?

**Comment by markmc** [comment]
> My only concern is whether we're not breaking API here.

See my comment. If we talk about this in terms of relaxing the "you may only use these 3 capabilities together" to "you may use 2 of these capabilities, without the other" ... it becomes more obvious to me that this isn't an API-breaking change. (i.e. changing `foo(bar, blaa, baz)` to `foo(bar, blaa, baz=None)` isn't an API break)

**Comment by stbenjam** [comment]
> I think we're missing a description of how exactly openshift-install will be used in an assisted installation. It's touched on in this enhancement and #376, but I think anyone not close to the subject would come away quite confused.

> Like, firstly an assisted install only ever uses "create ignition-configs" and not "create cluster". (I guess some would say this implies it's a UPI use case)

Perhaps we should explicitly call this out in this enhancement. The implied point of this enhancement is to facilitate the use of the `baremetal` platform for a UPI-like workflow. That is, allow you to generate manifests and ignitions without us actually requiring the fields needed for the automated bootstrap and master provisioning parts.

Deferring checking those fields is totally consistent with how other platforms do things. Even if we're doing IPI, Provisioning and Platform checks are clearly the right place for those validations; we made a mistake by validating **everything** in the installconfig validations. 

This is a bit of a contrived example but imagine a user wanting to do an IPI install, but she has a bunch of manifests and ignition customizations. It makes sense to let her do that on a development laptop, and then bring that output to the provisioning host to run `create cluster`. That works if we rearrange the validations to be run only where needed. It also allows the UPI-workflow assisted installer wants to do.

> My understanding of that intent is: (a) configuring the keepalived/haproxy API load balancer (which is IPI-ish), and (b) enable the baremetal-operator for day 2 machine API operations (usually only available for IPI). And not (c) "provision a libvirt VM with the bootstrap configuration, and a bunch of IPMI-controlled physical machines as control plane nodes using PXE" (definitely IPI).

Why are a and b separate? The way I understand things, we're not supporting separation: if you specify `baremetal` platform you get (a) and (b). We've deferred the epic to make Metal3 entirely optional.

> Related, AIUI the "configure the baremetal-operator" intent currently requires all of the fields in the metal3.io Provisioning? But we desire that to be a day 2 "set up PXE-based machine provisioning" thing? (That topic is touched on in openshift/machine-api-operator#560)

Well, that's changing. https://github.com/openshift/enhancements/pull/212 covers the optional provisioning network. It's sort of orthogonal to the assisted install case, even if they'll use it. It's valid even in IPI.

Although one note: all provisioning decisions on day 1 are permanent. We don't intend to make it possible to turn on a provisioning network later. If you install without it (either in assisted or IPI), you get Metal3 but can only do virtual media based provisioning and all the power management things. Trying to change the Provisioning configuration on day 2 is a separate epic and set of stories we could visit in the 4.7 time frame.

**Comment by stbenjam** [comment]
@markmc If we instead went to a capabilities-based approach, what would the interface look like?

**Comment by markmc** [comment]
> @markmc If we instead went to a capabilities-based approach, what would the interface look like?

Something like "the presence of field A implies capabilities 1 and 2, and field B implies capability 3" ?

(i.e. we want to make field B optional if you just want capability 1 and 2)

Maybe the key difference is the docs for this would describe how to use those capabilities individually rather than talking about which fields are validated by which phases

**Comment by markmc** [comment]
> > My understanding of that intent is: (a) configuring the keepalived/haproxy API load balancer (which is IPI-ish), and (b) enable the baremetal-operator for day 2 machine API operations (usually only available for IPI). And not (c) "provision a libvirt VM with the bootstrap configuration, and a bunch of IPMI-controlled physical machines as control plane nodes using PXE" (definitely IPI).
> 
> Why are a and b separate? The way I understand things, we're not supporting separation: if you specify `baremetal` platform you get (a) and (b). We've deferred the epic to make Metal3 entirely optional.

Right, but I could imagine a future enhancement making it possible to use (a) alone in UPI installs ... and then couldn't rely on the fields/phases association, so we'd have to switch then to thinking in terms of individual capabilities. Why not make that switch in approach now?

(I think I'm advocating for more of a change in documentation - this enhancement and user docs - rather than changing the mechanism)

**Comment by dhellmann** [comment]
> > > My understanding of that intent is: (a) configuring the keepalived/haproxy API load balancer (which is IPI-ish), and (b) enable the baremetal-operator for day 2 machine API operations (usually only available for IPI). And not (c) "provision a libvirt VM with the bootstrap configuration, and a bunch of IPMI-controlled physical machines as control plane nodes using PXE" (definitely IPI).
> > 
> > 
> > Why are a and b separate? The way I understand things, we're not supporting separation: if you specify `baremetal` platform you get (a) and (b). We've deferred the epic to make Metal3 entirely optional.
> 
> Right, but I could imagine a future enhancement making it possible to use (a) alone in UPI installs ... and then couldn't rely on the fields/phases association, so we'd have to switch then to thinking in terms of individual capabilities. Why not make that switch in approach now?
> 
> (I think I'm advocating for more of a change in documentation - this enhancement and user docs - rather than changing the mechanism)

How would the user express which capabilities they want?

**Comment by kirankt** [comment]
> > > > My understanding of that intent is: (a) configuring the keepalived/haproxy API load balancer (which is IPI-ish), and (b) enable the baremetal-operator for day 2 machine API operations (usually only available for IPI). And not (c) "provision a libvirt VM with the bootstrap configuration, and a bunch of IPMI-controlled physical machines as control plane nodes using PXE" (definitely IPI).
> > > 
> > > 
> > > Why are a and b separate? The way I understand things, we're not supporting separation: if you specify `baremetal` platform you get (a) and (b). We've deferred the epic to make Metal3 entirely optional.
> > 
> > 
> > Right, but I could imagine a future enhancement making it possible to use (a) alone in UPI installs ... and then couldn't rely on the fields/phases association, so we'd have to switch then to thinking in terms of individual capabilities. Why not make that switch in approach now?
> > (I think I'm advocating for more of a change in documentation - this enhancement and user docs - rather than changing the mechanism)
> 
> How would the user express which capabilities they want?

My guess is something similar to what @stbenjam proposed in https://github.com/openshift/enhancements/pull/386. Sort of choice in a IPI platform-wide profile, which enables/disables a whole set of related-features based on a said profile. Not sure about what features need to be turned on/off but couple things come to mind right away: validations (this proposal), provisioning network (@stbenjan's) proposal, etc.

Profiles could be:
Full deploy of IPI (default)
Day 2 deploy only (UPI, Assisted, ISO-based installs, etc)
...

**Comment by dhellmann** [comment]
@markmc I have incorporated the feedback you gave me earlier in the latest draft.

**Comment by dhellmann** [comment]
I haven't changed the title. Should I rename this to incorporate "user-provisioned" instead of focusing on the assisted installer?

**Comment by stbenjam** [comment]
> My guess is something similar to what @stbenjam proposed in #386. Sort of choice in a IPI platform-wide profile, which enables/disables a whole set of related-features based on a said profile.

At this point, I'd prefer not to combine the provisioning network profiles with other settings -- we may need some explicit indication later that an install as assisted, but I think this proposal covers what we need now.

**Comment by dhellmann** [comment]
> > My guess is something similar to what @stbenjam proposed in #386. Sort of choice in a IPI platform-wide profile, which enables/disables a whole set of related-features based on a said profile.
> 
> At this point, I'd prefer not to combine the provisioning network profiles with other settings -- we may need some explicit indication later that an install as assisted, but I think this proposal covers what we need now.

Yeah, I misunderstood what Mark was suggesting. After chatting with him it was clear that he meant the framing for this proposal was focused on the solution rather than the problem, and that thinking about the problem more broadly would result in a more useful problem statement and better context for reviewers not already familiar with the assisted installer. I've tried to incorporate his feedback into the latest draft (including stealing directly from some sample text he provided).

**Comment by markmc** [comment]
> > > My guess is something similar to what @stbenjam proposed in #386. Sort of choice in a IPI platform-wide profile, which enables/disables a whole set of related-features based on a said profile.
> > 
> > 
> > At this point, I'd prefer not to combine the provisioning network profiles with other settings -- we may need some explicit indication later that an install as assisted, but I think this proposal covers what we need now.
> 
> Yeah, I misunderstood what Mark was suggesting. After chatting with him it was clear that he meant the framing for this proposal was focused on the solution rather than the problem, and that thinking about the problem more broadly would result in a more useful problem statement and better context for reviewers not already familiar with the assisted installer. I've tried to incorporate his feedback into the latest draft (including stealing directly from some sample text he provided).

Thanks for your patience with my feedback. I'm happy with the proposal, and I hope the new framing of the problem helps capture where it's coming from at a higher level. :+1:

**Comment by stbenjam** [comment]
Are we all happy with this proposal then? Feel free to remove the hold when ready.

/lgtm
/hold

File: enhancements/installer/assisted-installer-bare-metal-validations.md [hunk 6]
```diff
@@ -0,0 +1,280 @@
...
+bare-metal user-provisioned infrastructure clusters, including those
+produced with the assisted installer, to include those
+installer-provisioned infrastructure capabilities.
+
+## Motivation
+
+The [connected-assisted-installer
+enhancement](./connected-assisted-installer.md) describes how an agent
+running on several bare-metal hosts receive a command from the
+assisted installer application to begin the installation of that
```
Comment (romfreiman) [inline enhancements/installer/assisted-installer-bare-metal-validations.md:46]: Personally, I dont like the term application. I would call it either assisted installer or assisted installer service

File: enhancements/installer/assisted-installer-bare-metal-validations.md [hunk 7]
```diff
@@ -0,0 +1,280 @@
...
+host. One of the first tasks that each host performs upon receiving
+this instruction is to download, from the assisted installer, the
+appropriate Ignition file for that host's role. These ignition files
+are generated by the assisted installer earlier in the process using
+the `openshift-install create ignition-configs` command.
+
+This installer workflow mirrors the [bare-metal user-provisioned
+infrastructure
+workflow](https://github.com/openshift/installer/blob/master/docs/user/metal/install_upi.md)
+in that the the OpenShift installer is used to generate assets for the
```
Comment (romfreiman) [inline enhancements/installer/assisted-installer-bare-metal-validations.md:56]: the the

File: enhancements/installer/assisted-installer-bare-metal-validations.md [hunk 8]
```diff
@@ -0,0 +1,280 @@
...
+  features on bare-metal user-provisioned infrastructure clusters
+- Avoid making any incompatible changes to the bare-metal
+  installer-provisioned infrastructure workflow
+
+### Non-Goals
+
+- Removing or relaxing any validation rules for the
+  installer-provisioned infrastructure installer.
+- Recapitulate the test plans for the assisted installer.
+- Recapitulate the [approach for making parts of the bare metal
```
Comment (romfreiman) [inline enhancements/installer/assisted-installer-bare-metal-validations.md:80]: [

File: enhancements/installer/assisted-installer-bare-metal-validations.md [hunk 9]
```diff
@@ -0,0 +1,280 @@
...
+The assisted installer does not need the `bootstrapProvisioningIP`,
+`provisioningNetworkCIDR`, `provisioningDHCPRange`, or
+`provisioningNetworkInterface` fields. The values in those fields are
+used to configure the provisioning tools used for the
+installer-provisioned infrastructure workflow, so those validation
+checks could move to a new function in the
+`pkg/types/baremetal/validation` package that can be invoked from
+[PlatformProvisionCheck.Generate()](https://github.com/openshift/installer/blob/master/pkg/asset/installconfig/platformprovisioncheck.go#L37).
+
+There are two validation rules for the `clusterOSImage` and
```
Comment (romfreiman) [inline enhancements/installer/assisted-installer-bare-metal-validations.md:162]: what are those used for? just for general knowledge

**Comment by romfreiman** [comment]
Looks good. I think we also discussed BMH annotations, but not 100% it's relevant to this enhancement

File: enhancements/installer/assisted-installer-bare-metal-validations.md [hunk 10]
```diff
@@ -0,0 +1,280 @@
...
+- Avoid making any incompatible changes to the bare-metal
+  installer-provisioned infrastructure workflow
+
+### Non-Goals
+
+- Removing or relaxing any validation rules for the
+  installer-provisioned infrastructure installer.
+- Recapitulate the test plans for the assisted installer.
+- Recapitulate the [approach for making parts of the bare metal
+  infrastructure automation optional](./baremetal-provisioning-optional.md).
```
Comment (n1r1) [inline enhancements/installer/assisted-installer-bare-metal-validations.md:81]: broken link? couldn't find that doc

Comment (n1r1) [reply enhancements/installer/assisted-installer-bare-metal-validations.md:81]: [Found it](https://github.com/openshift/enhancements/blob/master/enhancements/baremetal/baremetal-provisioning-optional.md). I guess the link will work once the enhancement will be merged.

File: enhancements/installer/assisted-installer-bare-metal-validations.md (hunk 9 above)
Comment (hardys) [reply enhancements/installer/assisted-installer-bare-metal-validations.md:162]: They are used to optionally specify a URL for the two RHCOS images needed for an IPI baremetal deployment - these are large and by default are downloaded every deploy from the internet, so this interface enables faster repeated testing and also deployment in disconnected environments.

File: enhancements/installer/assisted-installer-bare-metal-validations.md [hunk 11]
```diff
@@ -0,0 +1,280 @@
...
+
+##### Tech Preview -> GA 
+
+- More testing (upgrade, downgrade, scale)
+- Sufficient time for feedback
+- Available by default
+
+**For non-optional features moving to GA, the graduation criteria must include
+end to end tests.**
+
```
Comment (hardys) [inline enhancements/installer/assisted-installer-bare-metal-validations.md:240]: Some of the sections about graduation etc don't really apply here, can we just remove them?

**Comment by dhellmann** [comment]
I think this is ready to merge. We can address any other issues as follow-up patches.

/hold cancel

**Pull request merged**